### PR TITLE
Add alt text insertnig media

### PR DESCRIPTION
**Pull request by jarosik10**
for: https://github.com/openstax-poland/cnx-designer/issues/116

### CODE REVIEW

**Review by aiwenar — CHANGES_REQUESTED** [review]
This change will only affect media elements created using `insertFigure`, but not ones created using `insertSubfigure` or manually via `Transforms.insertNodes`. It also won't ensure that all `Media` elements have alt texts, as `AltText` can still be removed after `Media` is created. 

It will also insert a second `AltText` if `media` argument to `insertFigure` already is a `Media` node with an `AltText` child.

Finally, requiring that all `Media` elements have `AltText` is part of document schema, which in Slate is enforced with normalizations. Please at this as a normalization in `src/normalizations/figure.ts`.

However, it might still be worth including `AltText` in newly created figures as it would then save one round of normalizations. In such case, please do it in `itemToChildren`.

**Comment by jarosik10** [comment]
@aiwenar Do we want all `Media` elements include `alt`, even audio? This normalization makes some test fail right now.

File: src/normalize/figure.ts [hunk 1]
```diff
@@ -87,6 +87,16 @@ export default function normalizeFigure(editor: Editor, entry: NodeEntry): boole
...
             return true
         }
+
+        // Media elements must contain alt text.
+        if (node.children.length > 0 && !AltText.isAltText(node.children[node.children.length - 1])) {
+            const alt = {
+                type: 'media_alt',
+                children: [{ text: '' }]
+            }
+            Transforms.insertNodes(editor, alt, { at: [...path, node.children.length] })
```
Comment (aiwenar) [inline src/normalize/figure.ts:97]:
```suggestion
            Transforms.insertNodes(editor, {
                type: 'media_alt',
                children: [{ text: '' }]
            }, { at: [...path, node.children.length] })
```

File: src/transforms/figure.ts [hunk 2]
```diff
@@ -4,18 +4,26 @@
...
 export type MediaItem = Media | Omit<Audio | Image | Video, 'children'>
 
 function itemToChildren(media: MediaItem): Node[] {
+    const alt = {
+        type: 'media_alt',
+        children: [{ text: '' }]
+    }
+
     if (Media.isMedia(media)) {
+        if (!AltText.isAltText(media.children[media.children.length - 1])) {
```
Comment (aiwenar) [inline src/transforms/figure.ts:18]:
```suggestion
        if (!media.children.some(AltText.isAltText)) {
```

You shouldn't assume that existing `AltText` element will be last.

**Review by aiwenar — CHANGES_REQUESTED** [review]
> Do we want all `Media` elements include `alt`, even audio? This normalization makes some test fail right now.

If you look at [CNXML documentation for the `<media>` element](https://legacy.cnx.org/eip-help/media) you'll notice that the `alt` attribute is required, regardless of what media elements there are. I'd say that tests need to be adjusted.

**Review by aiwenar — APPROVED** [review]
Looks good now. Please squash commits and it will be good to merge.

**Comment by jarosik10** [comment]
@aiwenar should I fix the tests inside this PR as well? And how exactly should the commits be squashed?

**Comment by aiwenar** [comment]
> should I fix the tests inside this PR as well?

Yes, these are all failures related to this PR.

> And how exactly should the commits be squashed?

Such that the commit history reflects evolution of the project, rather than the back and forth between the committer and the reviewer.

**Comment by PiotrKozlowski** [comment]
@aiwenar Prośba o merge też tego pr jeżeli wszystko się zgadza

**Pull request merged**